solidity/contracts/deposit/DepositUtils.sol

```diff
@@ -448,30 +448,68 @@ library DepositUtils {
     /// @return             The amount in TBTC needed to redeem the deposit.
     function getOwnerRedemptionTbtcRequirement(DepositUtils.Deposit storage _d, address _redeemer) internal view returns(uint256) {
-        uint256 fee = signerFee(_d);
-        bool inCourtesy = _d.inCourtesyCall();
-        if(remainingTerm(_d) > 0 && !inCourtesy){
-            if(feeRebateTokenHolder(_d) != _redeemer) {
-                return fee;
-            }
-        }
-        uint256 contractTbtcBalance = _d.tbtcToken.balanceOf(address(this));
-        if(contractTbtcBalance < fee) {
-            return fee.sub(contractTbtcBalance);
-        }
-        return 0;
+        uint256 frtAdjust = frtAdjust(_d, _redeemer);
+        uint256 signerFeeAdjust = signerFeeAdjust(_d, _redeemer);
+        return frtAdjust.add(signerFeeAdjust);
     }
 
     /// @notice             Get TBTC amount required by redemption by a specified _redeemer.
     /// @dev                Will revert if redemption is not possible by msg.sender.
     /// @param _redeemer    The deposit redeemer.
     /// @return             The amount in TBTC needed to redeem the deposit.
     function getRedemptionTbtcRequirement(DepositUtils.Deposit storage _d, address _redeemer) internal view returns(uint256) {
+        uint256 mainCharge = mainCharge(_d, _redeemer);
+        uint256 frtAdjust = frtAdjust(_d, _redeemer);
+        uint256 signerFeeAdjust = signerFeeAdjust(_d, _redeemer);
+        return mainCharge.add(frtAdjust).add(signerFeeAdjust);
+    }
+
+    /// @notice             Get the base TBTC amount needed to redeem.
+    /// @param _redeemer    The deposit redeemer.
+    /// @return             The amount in TBTC.
+    function mainCharge(DepositUtils.Deposit storage _d, address _redeemer) internal view returns (uint256){
         bool inCourtesy = _d.inCourtesyCall();
-        if (depositOwner(_d) == _redeemer && !inCourtesy) {
-            return getOwnerRedemptionTbtcRequirement(_d, _redeemer);
+        bool ownerRedemption = depositOwner(_d) == _redeemer;
+        bool atTerm = remainingTerm(_d) == 0;
+
+        require(inCourtesy || ownerRedemption || atTerm, "Only TDT holder can redeem unless deposit is at-term or in COURTESY_CALL");
+
+        if (ownerRedemption) {
+            return 0;
         }
-        require(remainingTerm(_d) == 0 || inCourtesy, "Only TDT holder can redeem unless deposit is at-term or in COURTESY_CALL");
+
         return lotSizeTbtc(_d);
     }
+
+    /// @notice             Get the fee rebate amount needed to redeem.
+    /// @param _redeemer    The deposit redeemer.
+    /// @return             The amount in TBTC.
+    function frtAdjust(DepositUtils.Deposit storage _d, address _redeemer) internal view returns (uint256){
+        if(
+            _d.inCourtesyCall() ||
+            remainingTerm(_d) == 0 ||
+            feeRebateTokenHolder(_d) == address(0) ||
+            feeRebateTokenHolder(_d) == _redeemer
```

Comment:
`bool redeemerPaysFee` -> `bool redeemerPaysNoFee`
since no fee is owed if wither is true

Not sure if passing this as a param is really going to help clarify things, we just move the 2 lines to parent.

Reusing this bool on the signer-fee adjustment calculation makes this more palatable 

We would also still need to check if FRT exists since we can't rely on just `redeemerHoldsFrt`

Reply:
also:
`ownerRedemption` -> `tdtOwnerRedemption`
`redeemerHoldsFrt` -> `frtOwnerRedemption`